## Lambda functions

When working with data we often need to categorize our data differently, create new variables and do transformations on existing variables. Lambda functions make these types of transformations quick! 

Objectives of this lab: 
1. Intro you to lambda functions
2. Extra practice working with dataframes and transforming our data
3. Bonus! intro you to datetypes and datedeltas

Data set:   
The dataset comes from CouncilStat, which is used by many NYC Council district offices to enter and track constituent cases that can range from issues around affordable housing, to potholes and pedestrian safety. This dataset aggregates the information that individual staff have input. However, district staffs handle a wide range of complex issues. Each offices uses the program differently, and thus records cases, differently and so comparisons between accounts may be difficult. Not all offices use the program. For more info - http://labs.council.nyc/districts/data/

## Let's get started!
#### Import our libraries and our data

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("https://data.cityofnewyork.us/api/views/edai-dig6/rows.csv?accessType=DOWNLOAD&bom=true&query=select+*")

#### Let's get to know our data a bit

In [2]:
df.head()

UNIQUE_KEY ACCOUNT    OPENDATE         COMPLAINT_TYPE  \
0  NYCC44518150  NYCC44  05/03/2016         Transportation   
1  NYCC18502037  NYCC18  11/07/2016  Housing and Buildings   
2  NYCC38507613  NYCC38  05/05/2016        General Welfare   
3  NYCC31503037  NYCC31  11/23/2017         Youth Services   
4  NYCC40511790  NYCC40  11/07/2016                    NaN   

               DESCRIPTOR    ZIP   BOROUGH                 CITY COUNCIL_DIST  \
0     Taxi Fare Complaint  11230  Brooklyn             Brooklyn       NYCC44   
1             NYCHA Misc.  10472     Bronx                Bronx       NYCC18   
2  Child Abuse or Neglect    NaN  Brooklyn             Brooklyn          NaN   
3                     NaN  11413       NaN  Springfield Gardens         NYCC   
4                     NaN    NaN       NaN             New York          NaN   

  COMMUNITY_BOARD   CLOSEDATE  
0     12 Brooklyn  05/06/2016  
1        09 Bronx  11/07/2016  
2             NaN  05/05/2016  
3                         NaN  
4                         NaN

#### How many observations/rows are in our data? 

In [3]:
df.shape

(190634, 11)

#### We have 190,634 observations and 11 columns/variables. What are the variables? 

In [4]:
df.columns

Index([u'UNIQUE_KEY', u'ACCOUNT', u'OPENDATE', u'COMPLAINT_TYPE',
       u'DESCRIPTOR', u'ZIP', u'BOROUGH', u'CITY', u'COUNCIL_DIST',
       u'COMMUNITY_BOARD', u'CLOSEDATE'],
      dtype='object')

#### Great! Let's take a look at one of them "COMPLAINT_TYPE"

In [5]:
df["COMPLAINT_TYPE"].head(5)

0           Transportation
1    Housing and Buildings
2          General Welfare
3           Youth Services
4                      NaN
Name: COMPLAINT_TYPE, dtype: object

In [6]:
df["COMPLAINT_TYPE"].unique()

array(['Transportation', 'Housing and Buildings', 'General Welfare',
       'Youth Services', nan, 'Immigration', 'Finance',
       'Land Use and Zoning', 'Economy/Jobs', 'Consumer Affairs',
       'Cultural Affairs', 'Utilities', 'Sanitation', 'Education',
       'Governmental Operations', 'Legal Services', 'Parks',
       'Quality of Life', 'Environment', 'Public Safety', 'Health',
       'Aging', 'Human and Civil Rights', 'Civil Service and Labor',
       'Recovery and Resiliency', 'Consumer Complaints',
       'Veterans Affairs', 'NYPD - NEW YORK POLICE DE',
       'CITY CONSTRUCTION', 'CONTRACTS', 'OVERSIGHT AND INVESTIGATI',
       'Land Use', 'DOCUMENTATION', 'Select One', 'PENSION FORMS',
       'Government', 'JURY DUTY', 'Please choose an issue...', 'MEETING',
       'SPAM MAIL', 'SCRIE', 'DISPLACEMENT', 'PATRIOTISM', 'NYCHA', 'DWI',
       'DHS', 'EVNT', 'O AND I REQ', 'INT REQ',
       '@getxlate.udf_code_genera', '@getxlate.Udf_Code_Genera',
       'CRIMINAL JUSTICE', 'FIRE

#### Hmmm... some of these categories are in sentence case and some are in all caps. Let's clean this up a bit. 
One way to do this is write a named function (below it's called upper) to transform all the names to uppercase. That paired with .apply ([check out the docs here](href=https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html) which allows us to "apply" a function to every row in our dataframe, allows us to transform our data. Let's give it try

In [7]:
#Example of a normal function that turns a string in lower/sentence case to upper case 
def upper(x):
    return str(x).upper()

df["COMPLAINT_TYPE"] = df["COMPLAINT_TYPE"].apply(upper)

In [8]:
#Check to ensure the function worked. Yes!
df["COMPLAINT_TYPE"].unique()

array(['TRANSPORTATION', 'HOUSING AND BUILDINGS', 'GENERAL WELFARE',
       'YOUTH SERVICES', 'NAN', 'IMMIGRATION', 'FINANCE',
       'LAND USE AND ZONING', 'ECONOMY/JOBS', 'CONSUMER AFFAIRS',
       'CULTURAL AFFAIRS', 'UTILITIES', 'SANITATION', 'EDUCATION',
       'GOVERNMENTAL OPERATIONS', 'LEGAL SERVICES', 'PARKS',
       'QUALITY OF LIFE', 'ENVIRONMENT', 'PUBLIC SAFETY', 'HEALTH',
       'AGING', 'HUMAN AND CIVIL RIGHTS', 'CIVIL SERVICE AND LABOR',
       'RECOVERY AND RESILIENCY', 'CONSUMER COMPLAINTS',
       'VETERANS AFFAIRS', 'NYPD - NEW YORK POLICE DE',
       'CITY CONSTRUCTION', 'CONTRACTS', 'OVERSIGHT AND INVESTIGATI',
       'LAND USE', 'DOCUMENTATION', 'SELECT ONE', 'PENSION FORMS',
       'GOVERNMENT', 'JURY DUTY', 'PLEASE CHOOSE AN ISSUE...', 'MEETING',
       'SPAM MAIL', 'SCRIE', 'DISPLACEMENT', 'PATRIOTISM', 'NYCHA', 'DWI',
       'DHS', 'EVNT', 'O AND I REQ', 'INT REQ',
       '@GETXLATE.UDF_CODE_GENERA', 'CRIMINAL JUSTICE',
       'FIRE AND EMERGENCY MANAGE'], dt

#### A lambda function (or an anonymous or unnamed function) can do the same thing BUT is faster to type and handy for quick changes

In [10]:
#Example of a lambda function, e.g., do the same thing in one line (this time lower case)
df["COMPLAINT_TYPE_LOWER"] = df["COMPLAINT_TYPE"].apply(lambda x: str(x).lower())

#### Let's walk through this by comparing it to our function "upper" above

Starting with **lambda x: str(x).lower()**

lambda indicates we're using a lambda function x is our we're going to refer to our parameter in this case. It's equivalent to x in our upper function above. Just like in a regular function you can all this whatever you want.

str(x).lower() is equivalent to "return str(x).upper()" in the upper function. 

the rest .apply etc. allows us to apply the function to each row and create a new column called "COMPLAINT_TYPE_LOWER"

In [11]:
#Check to ensure the function worked. Yes!
df[["COMPLAINT_TYPE", "COMPLAINT_TYPE_LOWER"]].head(5)

COMPLAINT_TYPE   COMPLAINT_TYPE_LOWER
0         TRANSPORTATION         transportation
1  HOUSING AND BUILDINGS  housing and buildings
2        GENERAL WELFARE        general welfare
3         YOUTH SERVICES         youth services
4                    NAN                    nan

#### Try writing a lambda function that changes the values in City to uppercase

In [11]:
df["CITY"].head(10)

0               Brooklyn
1                  Bronx
2               Brooklyn
3    Springfield Gardens
4               New York
5               Brooklyn
6               Elmhurst
7               New York
8                    NaN
9               New York
Name: CITY, dtype: object

In [12]:
#write your function here

df["CITY"] = df["CITY"].apply(lambda x: str(x).upper())


In [13]:
#check our work
df["CITY"].head(10)

0               BROOKLYN
1                  BRONX
2               BROOKLYN
3    SPRINGFIELD GARDENS
4               NEW YORK
5                  10304
6               BROOKLYN
7               ELMHURST
8               NEW YORK
9                    NAN
Name: CITY, dtype: object

### Great! Let's try it with another function
Now let's use lambda to convert OPENDATE and CLOSEDDATE to "datetime" objects


##### This is how you would convert a column to a "datetime" object normally. Let's try it
df["OPENDATE"] = pd.to_datetime(df["OPENDATE"])

In [14]:
# note this takes a little while to run
print "Starting type:", df["OPENDATE"].dtype
df["OPENDATE"] = pd.to_datetime(df["OPENDATE"])
print "After conversion type:", df["OPENDATE"].dtype

Starting type: object
After conversion type: datetime64[ns]


#### How would you do it with a lambda function on CLOSEDATE?

In [15]:
print "Starting type:", df["CLOSEDATE"].dtype
df["CLOSEDATE"] = df["CLOSEDATE"].apply(lambda x: pd.to_datetime(x))
print "After conversion type:", df["CLOSEDATE"].dtype

Starting type: object
After conversion type: datetime64[ns]


### Great! Now let's say we wanted to group certain complaint types together.

Specifically, we want to divide the complaint types into three categories:  
       1. Government  
       2. Specific Group 
       3. Everything else/Other  


In [8]:
#Let's review what we have now
df["COMPLAINT_TYPE"].unique()

array(['Transportation', 'Housing and Buildings', 'General Welfare',
       'Youth Services', nan, 'Immigration', 'Finance',
       'Land Use and Zoning', 'Economy/Jobs', 'Consumer Affairs',
       'Cultural Affairs', 'Utilities', 'Sanitation', 'Education',
       'Governmental Operations', 'Legal Services', 'Parks',
       'Quality of Life', 'Public Safety', 'Environment', 'Health',
       'Aging', 'Human and Civil Rights', 'Civil Service and Labor',
       'Recovery and Resiliency', 'Consumer Complaints',
       'Veterans Affairs', 'NYPD - NEW YORK POLICE DE',
       'CITY CONSTRUCTION', 'Land Use', 'DOCUMENTATION', 'Select One',
       'NYCHA', 'PENSION FORMS', 'Government', 'JURY DUTY',
       'Please choose an issue...', 'MEETING', 'SPAM MAIL', 'SCRIE',
       'DISPLACEMENT', 'DHS', 'PATRIOTISM', 'LEFT TURN SIGNAL', 'DWI',
       'EVNT', 'O AND I REQ', 'INT REQ'], dtype=object)

#### Let's write a (named) function for that create the following groups:  
* "GOVERNMENTAL OPERATIONS", "GOVERNMENT" and "NYPD - NEW YORK POLICE DE" as Government-Related
* "IMMIGRATION", "AGING", "YOUTH SERVICES" and "VETERANS AFFAIRS" as Specific_Group  
* everything else as Other 

In [16]:
def categories(x):
    if (x == "GOVERNMENTAL OPERATIONS") or (x == "GOVERNMENT") or (x == "NYPD - NEW YORK POLICE DE"):
        return "Government-Related"
    elif (x == "IMMIGRATION") or (x =="AGING") or (x == "YOUTH SERVICES") or (x == "VETERANS AFFAIRS"):
        return "Specific_Group"
    else:
        return "Other"



#### now let's create a new column called CATEGORY and classify each complaint type. axis=1 will allow us to apply this to the row. [Learn more](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html


In [17]:
df["CATEGORY"] = df.apply(lambda x: categories(x["COMPLAINT_TYPE"]), axis=1)

In [18]:
#Check to make sure function worked. Great!
df[["COMPLAINT_TYPE","CATEGORY"]].head(10)

COMPLAINT_TYPE        CATEGORY
0         TRANSPORTATION           Other
1  HOUSING AND BUILDINGS           Other
2        GENERAL WELFARE           Other
3         YOUTH SERVICES  Specific_Group
4                    NAN           Other
5        GENERAL WELFARE           Other
6            IMMIGRATION  Specific_Group
7                FINANCE           Other
8    LAND USE AND ZONING           Other
9    LAND USE AND ZONING           Other

#### Whoa there! Did we just combine a named function (categories) and a lambda function!?! 
Yes! This will save you from so many nested for loops.... 

#### Let's walk through it!
** lambda x: categories(x["COMPLAINT_TYPE"])**

lambda x: **//that's the same as before**

Next we're going to call the function categories that takes a parameter x   
We're going to read in x["COMPLAINT_TYPE] for parameter x. 
What is x["COMPLAINT_TYPE"] exactly??

#### Let's add some print statements to see it in action with subset of our data

In [19]:
def categories2(x):
    print x
    if (x == "GOVERNMENTAL OPERATIONS") or (x == "GOVERNMENT") or (x == "NYPD - NEW YORK POLICE DE"):
        return "Government-Related"
    elif (x == "IMMIGRATION") or (x =="AGING") or (x == "YOUTH SERVICES") or (x == "VETERANS AFFAIRS"):
        return "Specific_Group"
    else:
        return "Other"



In [20]:
#creating a subset 
selection = df.sample(15, random_state=20)
print "number of sampeles:", len(selection)
selection.head()

number of sampeles: 15


UNIQUE_KEY ACCOUNT   OPENDATE         COMPLAINT_TYPE  \
63087   NYCC38507904  NYCC38 2016-08-10    LAND USE AND ZONING   
90316   NYCC08503461  NYCC08 2015-06-30  HOUSING AND BUILDINGS   
106436  NYCC34514115  NYCC34 2015-01-26                FINANCE   
28834   NYCC10507998  NYCC10 2016-12-13                    NAN   
58704   NYCC38508088  NYCC38 2016-09-28                  PARKS   

                   DESCRIPTOR    ZIP    BOROUGH      CITY COUNCIL_DIST  \
63087                  Zoning  11220   Brooklyn  BROOKLYN       NYCC38   
90316   NYCHA New Application  10454  Manhattan     BRONX       NYCC08   
106436   Tax Preparation Assi  11249   Brooklyn  BROOKLYN       NYCC33   
28834                     NaN  10453      Bronx     BRONX       NYCC14   
58704                     NaN   1122   Brooklyn  BROOKLYN       NYCC38   

       COMMUNITY_BOARD  CLOSEDATE   COMPLAINT_TYPE_LOWER CATEGORY  
63087      07 Brooklyn        NaT    land use and zoning    Other  
90316     01 Manhattan 2015-06-30  housing and buildings    Other  
106436     01 Brooklyn 2015-01-26                finance    Other  
28834         05 Bronx        NaT                    nan    Other  
58704      07 Brooklyn        NaT                  parks    Other

In [21]:
#running the same lambda funtion on the subset, but printing x
selection["CATEGORY"] = selection.apply(lambda x: categories2(x["COMPLAINT_TYPE"]), axis=1)

LAND USE AND ZONING
HOUSING AND BUILDINGS
FINANCE
NAN
PARKS
LAND USE AND ZONING
PARKS
FINANCE
HOUSING AND BUILDINGS
SANITATION
HOUSING AND BUILDINGS
GENERAL WELFARE
HOUSING AND BUILDINGS
PARKS
SANITATION


#### Can you write in a sentence what this is doing?

Answer: It's taking our subset, and applying the "categories2" function to it, via lambda function

In [22]:
#check your work
selection[["COMPLAINT_TYPE", "CATEGORY"]]

COMPLAINT_TYPE CATEGORY
63087     LAND USE AND ZONING    Other
90316   HOUSING AND BUILDINGS    Other
106436                FINANCE    Other
28834                     NAN    Other
58704                   PARKS    Other
74201     LAND USE AND ZONING    Other
147582                  PARKS    Other
109214                FINANCE    Other
163245  HOUSING AND BUILDINGS    Other
32181              SANITATION    Other
181524  HOUSING AND BUILDINGS    Other
95665         GENERAL WELFARE    Other
70301   HOUSING AND BUILDINGS    Other
127809                  PARKS    Other
7249               SANITATION    Other

### Now it's your turn! Create your own function using the "Descriptor" column.
Divide them into 5 categories: 
    1. Medicaid or Food Stamps  
    2. Immigration or Green Card or Literary Services for Immigrants  
    3. Seeking an affordable housing/apartment, Landlord/Mangaging Agent Issues, HPD Housing Lottery, 
    4. Tree Planting/Pruning, Street Resurfacing, Construction General
    5. Everything else/other

Name them: 
    1. Benefits
    2. Immigration, 
    3. Housing
    4. Outside_Amenities
    5. Other

In [25]:
def categorizor(x):
    if (x == "Medicaid") or (x == "Food Stamps"):
        return "Benefits"
    elif (x == "Immigration") or (x == "Green Card") or (x == "Literary Services for Immigrants"):
        return "Immigration"
    elif (x == "Seeking an affordable housing/apartment") or (x == "Landlord/Managing Agent Issues") or (x == "HPD Housing Lottery"):
        return "Housing"
    elif (x == "Tree Planting/Pruning") or (x == "Street Resurfacing") or (x == "Construction General"):
        return "Outside_Amenities"
    else:
        return "Other"

df["Categorizor"] = df.apply(lambda x: categorizor(x["DESCRIPTOR"]), axis = 1)



In [27]:
df["Categorizor"].value_counts()

Other                163998
Housing                9483
Outside_Amenities      7496
Immigration            5045
Benefits               4612
Name: Categorizor, dtype: int64

In [28]:
#Use .value_counts() and .unique() methods to help with spelling errors etc. 
df["DESCRIPTOR"].value_counts()

U.S. Citizenship                                6742
Tax Preparation Assi                            5543
Seeking an affordable housing/apartment         4885
Street Resurfacing                              4062
Lease Issue                                     3750
Landlord/Managing Agent Issues                  3344
Food Stamps                                     3235
Misc (Wills/Court Docs/etc.)                    2917
NYCHA Building Maintenance                      2579
Eviction                                        2295
SCRIE                                           2286
NYCHA Misc.                                     2020
NYCHA New Application                           1990
Tree Planting/Pruning                           1919
Green Card                                      1838
Libraries                                       1810
Section 8                                       1757
Traffic Signals                                 1741
Literary Services for Immigrants              

In [23]:
#check your work!
df["DESCRIPTOR"].value_counts()

U.S. Citizenship                                6406
Tax Preparation Assi                            5286
Seeking an affordable housing/apartment         4457
Food Stamps                                     3037
Street Resurfacing                              2986
Landlord/Managing Agent Issues                  2939
Lease Issue                                     2630
Misc (Wills/Court Docs/etc.)                    2585
NYCHA Building Maintenance                      2156
Eviction                                        2067
SCRIE                                           1995
NYCHA New Application                           1850
Libraries                                       1790
Green Card                                      1742
NYCHA Misc.                                     1680
Literary Services for Immigrants                1652
Tree Planting/Pruning                           1643
Section 8                                       1622
Immigration                                   

### Great job!  Below is a bonus excerise using data times! 
Create new column, "Complaints_Duration", 
which subtracts"Closed Date" column from "Open Date" to see how many days it takes to get complaints resolved

In [42]:
df["COMPLAINTS_DURATION"] = df["CLOSEDATE"] - df["OPENDATE"]
df["COMPLAINTS_DURATION"].head()

0   3 days
1   0 days
2   0 days
3      NaT
4      NaT
Name: COMPLAINTS_DURATION, dtype: timedelta64[ns]

In [43]:
df["COMPLAINTS_DURATION"].describe()

count                      164678
mean      50 days 07:42:00.424100
std      122 days 18:04:12.749480
min               0 days 00:00:00
25%               0 days 00:00:00
50%               5 days 00:00:00
75%              41 days 00:00:00
max            1327 days 00:00:00
Name: COMPLAINTS_DURATION, dtype: object

In [44]:
pd.to_timedelta(df["COMPLAINTS_DURATION"])
df["COMPLAINTS_DURATION"].dtype

dtype('<m8[ns]')

#### learn more about time deltas here! 
https://pandas.pydata.org/pandas-docs/stable/timedeltas.html

In [33]:
# df["COMPLAINTS_DURATION"] = df["COMPLAINTS_DURATION"]
df["COMPLAINTS_DURATION"].head()

0   3 days
1   0 days
2   0 days
3      NaT
4      NaT
Name: COMPLAINTS_DURATION, dtype: timedelta64[ns]

In [52]:
#Now create a column using a lambda function which indicates "duration ranges", 0-30 days, 31-60 days, 60-90 days, 90+ days

def duration_ranges(x):
    if (x < 30):
        return "One_Month"
    elif (30 < x < 60):
        return "Two_Months"
    elif (60 < x < 90):
        return "Three_Months"
    elif (90 < x):
        return "More than Three Months"
    else:
        return "Unknown"
    
    
    


In [53]:
df["Duration_Ranges"] = df.apply(lambda x: duration_ranges(x["COMPLAINTS_DURATION"].days), axis=1)

In [54]:
df[["COMPLAINTS_DURATION","Duration_Ranges"]].head(10)

COMPLAINTS_DURATION         Duration_Ranges
0              3 days               One_Month
1              0 days               One_Month
2              0 days               One_Month
3                 NaT                 Unknown
4                 NaT                 Unknown
5                 NaT                 Unknown
6              0 days               One_Month
7            291 days  More than Three Months
8              0 days               One_Month
9              3 days               One_Month

### Note that the function is classifying "N/A's" as "More Than Three Months."
#### Bonus! Fix that! 

### Learn more about lambda functions here:
https://community.modeanalytics.com/python/tutorial/pandas-groupby-and-python-lambda-functions/